lib/charms/ip_router_interface/v0/ip_router_interface.py

```diff
@@ -27,6 +136,344 @@
+    for existing_network_list in existing_routing_table.values():
+        for existing_network in existing_network_list:
+            old_subnet = IPv4Network(existing_network["network"])
+            new_subnet = IPv4Network(network_request["network"])
+
+            if old_subnet.subnet_of(new_subnet) or old_subnet.supernet_of(new_subnet):
+                raise ValueError("This network has been defined in a previous entry.")
+
+
+def _network_name_taken(name: str, relations: List[Relation]):
+    count = 0
+    for relation in relations:
+        if name == relation.data[relation.app].get("network-name"):
+            count += 1
+
+    if count > 1:
+        logger.error(
+            "There are multiple relations with the name (%s). Please change one or provide a custom network name.",
+            name,
+        )
+        return True
+    return False
+
+
+class RouterProvides(Object):
+    """This class is used to manage the routing table of the router provider.
+
+    It's capabilities are to:
+    * Manage the routing table in the charm itself, by adding and removing
+    new network and route requests by integrated units,
+    * Synchronize the databags of all requiring units with the router table of the
+    provider charm.
+
+    Attributes:
+        charm:
+            The Charm object that instantiates this class.
+        relationship_name:
+            The name used for the relationship implementing the ip-router interface
+            "ip-router" by default.
+    """
+
+    on = RouterProviderCharmEvents()
+
+    def __init__(self, charm: CharmBase, relationship_name: str = "ip-router"):
+        super().__init__(charm, relationship_name)
+        self.charm = charm
+        self.relationship_name = relationship_name
+        self.framework.observe(
+            charm.on[relationship_name].relation_changed, self._router_relation_changed
+        )
+        self.framework.observe(
+            charm.on[relationship_name].relation_departed, self._router_relation_departed
+        )
+
+    def _router_relation_changed(self, event: RelationChangedEvent):
+        """Resync the databags since there could have been a change in networks."""
+        if not self.charm.unit.is_leader():
+            return
+        self._sync_routing_tables()
+        new_table = self.get_flattened_routing_table()
+        self.on.routing_table_updated.emit({"networks": new_table})
+
+    def _router_relation_departed(self, event: RelationDepartedEvent):
+        """If an application has completely departed the relation, remove it
+        from the routing table.
+        """
+        if not self.charm.unit.is_leader():
+            return
+        self._sync_routing_tables()
+
+    def get_routing_table(self) -> RoutingTable:
+        """Build the routing table from all of the related databags. Relations
+        that don't have missing or invalid network requests will be ignored.
+        """
+        logger.debug("Routing Table generation starting.")
```

Comment:
Not sure of the value is this logging, try to log after actions have been taken

Reply:
I personally like logging before starting a long process because it gives a clear boundary from where errors could have originated from, but it's not a big deal. Removed.